pkg/kademlia/workers.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package kademlia
+
+import (
+	"container/heap"
+	"context"
+	"log"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/zeebo/errs"
+	"storj.io/storj/pkg/dht"
+	"storj.io/storj/pkg/node"
+	"storj.io/storj/pkg/pb"
+)
+
+var (
+	// WorkerError is the class of errors for the worker struct
+	WorkerError = errs.Class("worker error")
+	// default timeout is the minimum timeout for worker cancellation
+	// 250ms was the minimum value allowing current workers to finish work
+	// before returning
+	defaultTimeout = 250 * time.Millisecond
+)
+
+type worker struct {
+	contacted      map[string]bool
+	pq             PriorityQueue
+	mu             *sync.Mutex
+	maxResponse    time.Duration
+	cancel         context.CancelFunc
+	nodeClient     node.Client
+	find           dht.NodeID
+	workInProgress int
+	k              int
+}
+
+func newWorker(ctx context.Context, rt *RoutingTable, nodes []*pb.Node, nc node.Client, target dht.NodeID, k int) *worker {
+	t := new(big.Int).SetBytes(target.Bytes())
+
+	pq := func(nodes []*pb.Node) PriorityQueue {
+		pq := make(PriorityQueue, len(nodes))
+		for i, node := range nodes {
+			bnode := new(big.Int).SetBytes([]byte(node.GetId()))
+			pq[i] = &Item{
+				value:    node,
+				priority: new(big.Int).Xor(t, bnode),
+				index:    i,
+			}
+
+		}
+		heap.Init(&pq)
+
+		return pq
+	}(nodes)
+
+	return &worker{
+		contacted:      map[string]bool{},
+		pq:             pq,
+		mu:             &sync.Mutex{},
+		maxResponse:    0 * time.Millisecond,
+		nodeClient:     nc,
+		find:           target,
+		workInProgress: 0,
+		k:              k,
+	}
+}
+
+// create x workers
+// have a worker that gets work off the queue
+// send that work on a channel
+// have workers get work available off channel
+// after queue is empty and no work is in progress, close channel.
+
+func (w *worker) work(ctx context.Context, ch chan *pb.Node) {
+	// grab uncontacted node from working set
+	// change status to inprogress
+	// ask node for target
+	// if node has target cancel ctx and send node
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case n := <-ch:
+			// network lookup for nodes
+			nodes := w.lookup(ctx, n)
+			// update our priority queue
+			w.update(nodes)
+		}
+	}
+}
+
+func (w *worker) getWork(ctx context.Context, ch chan *pb.Node) {
+	for {
+		if ctx.Err() != nil {
+			return
+		}
+
+		w.mu.Lock()
+		if w.pq.Len() <= 0 && w.workInProgress <= 0 {
+			w.mu.Unlock()
+			timeout := defaultTimeout
+			if timeout < (2 * w.maxResponse) {
+				timeout = 2 * w.maxResponse
+			}
+
+			time.AfterFunc(timeout, w.cancel)
+			return
+		}
+
+		if w.pq.Len() <= 0 {
+			w.mu.Unlock()
+			// if there is nothing left to get off the queue
+			// and the work-in-progress is not empty
+			// let's wait a bit for the workers to populate the queue
+			time.Sleep(50 * time.Millisecond)
```

Comment:
Concurrent synchronized code shouldn't have sleeps. This usually indicates that there is a logical bug somewhere. If there is no bug, then the sleep can be removed.

There are uses where random sleep is used for avoiding thundering herd, but this doesn't look like it.

Reply:
Maybe it's a bug? What I was experiencing in testing was the worker that unlocked, would get the lock before the others. So I introduced a small sleep to give them a chance to grab a lock. You have any insights ?

Reply:
ok after talking via Slack, I've created https://storjlabs.atlassian.net/browse/V3-438 to adjust the worker pattern.